README.md
R pacakge: ncarbe
================
Sungpil Han
2018-06-07

<img src="https://shanmdphd.github.io/hex/ncarbe.png" height="200" alt="ncarbe R package" align="right" />

`NonCompart` 패키지를 사용해서 2x2 생동성시험의 분산분석 등을 손쉽게 해주는 패키지입니다. 지원하는 계산은 다음과 같습니다.

<table style="width:100%;">
<colgroup>
<col width="4%" />
<col width="95%" />
</colgroup>
<thead>
<tr class="header">
<th>parameters</th>
<th>statistics</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>AUClast</code></td>
<td>Analysis of variance, between and within subject variability, least square means, confidence interval, and sample size</td>
</tr>
<tr class="even">
<td><code>Cmax</code></td>
<td>Analysis of variance, between and within subject variability, least square means, confidence interval, and sample size</td>
</tr>
<tr class="odd">
<td><code>Tmax</code></td>
<td>Wilcoxon Signed-Rank Test and Hodges-Lehmann Estimate</td>
</tr>
</tbody>
</table>

설치
====

``` r
install.packages('devtools')
devtools::install_github('asancpt/ncarbe')
```

실행
====

예제 데이타 (시뮬레이션 데이타)
-------------------------------

2x2 생동성시험 디자인으로 36명의 데이타를 생성하여 시뮬레이션 한 것입니다. (배균섭 교수님 `pk1coma.R` 사용)

``` r
file <- system.file('example', 'beConc.csv', package = 'ncarbe')
concData <- read.csv(file, as.is = TRUE)
head(concData)
```

    ##   SUBJ GRP PRD TRT nTIME TIME   CONC
    ## 1    1  RT   1   R  0.00 0.02  63.42
    ## 2    1  RT   1   R  0.25 0.24 432.76
    ## 3    1  RT   1   R  0.50 0.51 622.88
    ## 4    1  RT   1   R  0.75 0.80 809.93
    ## 5    1  RT   1   R  1.00 1.02 824.34
    ## 6    1  RT   1   R  2.00 2.04 602.22

자료 탐색
---------

그림을 그려보면 다음과 같습니다.

``` r
library(ggplot2)
library(dplyr)

concData %>% 
  ggplot(aes(x = TIME, y = CONC, group = TRT, color = TRT)) +
  geom_line(size = 0.5, alpha = 0.8) +
  geom_point(size = 1, alpha = 0.5) +
  xlab('Time (hr)') +
  ylab('Concentrationg (ng/mL)') +
  facet_wrap( ~ SUBJ, ncol = 4)
```

![](README_files/figure-markdown_github/conctimeplot-1.png)

생동성 분석
-----------

`beNCA()` 함수를 사용하여 분석한 결과입니다.

### method = 'kbe'

``` r
library(ncarbe)
beNCA(concData)     # method = 'kbe'
```

    ## 
    ## 
    ## [AUClast]
    ## 
    ## $`Analysis of Variance`
    ##                        SS DF        MS        F           p
    ## SUBJECT        35.7817072 35 1.0223345 1.854676 0.037320327
    ## GROUP           1.1390308  1 1.1390308 1.117900 0.297825788
    ## SUBJECT(GROUP) 34.6426764 34 1.0189022 1.848449 0.038831903
    ## PERIOD          4.1989873  1 4.1989873 7.617625 0.009244748
    ## DRUG            0.7021705  1 0.7021705 1.273848 0.266947114
    ## ERROR          18.7414811 34 0.5512200       NA          NA
    ## TOTAL          58.9756619 71        NA       NA          NA
    ## 
    ## $`Between and Within Subject Variability`
    ##                                 Between Subject Within Subject
    ## Variance Estimate                     0.2338411        0.55122
    ## Coefficient of Variation, CV(%)      51.3267693       85.75365
    ## 
    ## $`Least Square Means`
    ##                 Reference Drug Test Drug
    ## Geometric Means       31.20441   25.5401
    ## 
    ## $`90% Confidence Interval`
    ##                  Lower Limit Point Estimate Upper Limit
    ## 90% CI for Ratio   0.6062815      0.8184771     1.10494
    ## 
    ## $`Sample Size`
    ##                       True Ratio=1 True Ratio=Point Estimate
    ## 80% Power Sample Size           96                      6538
    ## 
    ## 
    ## 
    ## [Cmax]

    ## Warning in sqrt(exp(sig2b) - 1): NaN이 생성되었습니다

    ## $`Analysis of Variance`
    ##                        SS DF        MS         F          p
    ## SUBJECT        22.1199814 35 0.6319995 0.6475704 0.89703843
    ## GROUP           0.2047080  1 0.2047080 0.3175901 0.57675784
    ## SUBJECT(GROUP) 21.9152734 34 0.6445669 0.6604475 0.88422171
    ## PERIOD          3.6728929  1 3.6728929 3.7633843 0.06071806
    ## DRUG            0.1136911  1 0.1136911 0.1164922 0.73497096
    ## ERROR          33.1824629 34 0.9759548        NA         NA
    ## TOTAL          58.9756619 71        NA        NA         NA
    ## 
    ## $`Between and Within Subject Variability`
    ##                                 Between Subject Within Subject
    ## Variance Estimate                     -0.165694      0.9759548
    ## Coefficient of Variation, CV(%)             NaN    128.5962569
    ## 
    ## $`Least Square Means`
    ##                 Reference Drug Test Drug
    ## Geometric Means       26.30688  28.51506
    ## 
    ## $`90% Confidence Interval`
    ##                  Lower Limit Point Estimate Upper Limit
    ## 90% CI for Ratio   0.7270811       1.083939    1.615946
    ## 
    ## $`Sample Size`
    ##                       True Ratio=1 True Ratio=Point Estimate
    ## 80% Power Sample Size          169                       298
    ## 
    ## 
    ## 
    ## [Tmax]
    ## 
    ## $`Wilcoxon Signed-Rank Test`
    ##   p-value 
    ## 0.3286283 
    ## 
    ## $`Hodges-Lehmann Estimate`
    ##                            Lower Limit Point Estimate Upper Limit
    ## 90% Confidence Interval       -2.00000         1.0000      4.0000
    ## 90% Confidence Interval(%)    87.11261       106.4437    125.7748

### method = 'nlme'

``` r
beNCA(concData, method = 'nlme')
```

    ## $Confidence.Interval
    ## # A tibble: 2 x 4
    ##   parameter `Lower limit of 90% CI` `T/R ratio` `Upper limit of 90% CI`
    ##   <chr>                       <dbl>       <dbl>                   <dbl>
    ## 1 Cmax                        0.753       1.08                     1.56
    ## 2 AUClast                     0.606       0.818                    1.10

Reference
=========

-   <p>
    Bae K (2018). <em>NonCompart: Noncompartmental Analysis for Pharmacokinetic Data</em>. R package version 0.4.2, <a href="https://CRAN.R-project.org/package=NonCompart">https://CRAN.R-project.org/package=NonCompart</a>.
    </p>
-   <p>
    Bae K, Han S (2018). <em>ncarbe: Perform BE Test Using NonCompart Package</em>. R package version 0.1.1.
    </p>
-   <p>
    Pinheiro J, Bates D, DebRoy S, Sarkar D, R Core Team (2018). <em>nlme: Linear and Nonlinear Mixed Effects Models</em>. R package version 3.1-137, <a href="https://CRAN.R-project.org/package=nlme">https://CRAN.R-project.org/package=nlme</a>.
    </p>

devtools::session\_info()
=========================

``` r
devtools::session_info()
```

    ## Session info -------------------------------------------------------------

    ##  setting  value                       
    ##  version  R version 3.5.0 (2018-04-23)
    ##  system   x86_64, darwin17.5.0        
    ##  ui       unknown                     
    ##  language (EN)                        
    ##  collate  C                           
    ##  tz       Asia/Seoul                  
    ##  date     2018-06-07

    ## Packages -----------------------------------------------------------------

    ##  package    * version date       source        
    ##  assertthat   0.2.0   2017-04-11 CRAN (R 3.5.0)
    ##  backports    1.1.2   2017-12-13 CRAN (R 3.5.0)
    ##  base       * 3.5.0   2018-05-02 local         
    ##  bindr        0.1.1   2018-03-13 CRAN (R 3.5.0)
    ##  bindrcpp   * 0.2.2   2018-03-29 CRAN (R 3.5.0)
    ##  cli          1.0.0   2017-11-05 CRAN (R 3.5.0)
    ##  colorspace   1.3-2   2016-12-14 CRAN (R 3.5.0)
    ##  compiler     3.5.0   2018-05-02 local         
    ##  crayon       1.3.4   2017-09-16 CRAN (R 3.5.0)
    ##  datasets   * 3.5.0   2018-05-02 local         
    ##  devtools     1.13.5  2018-02-18 CRAN (R 3.5.0)
    ##  digest       0.6.15  2018-01-28 CRAN (R 3.5.0)
    ##  dplyr      * 0.7.5   2018-05-19 CRAN (R 3.5.0)
    ##  evaluate     0.10.1  2017-06-24 CRAN (R 3.5.0)
    ##  ggplot2    * 2.2.1   2016-12-30 CRAN (R 3.5.0)
    ##  glue         1.2.0   2017-10-29 CRAN (R 3.5.0)
    ##  graphics   * 3.5.0   2018-05-02 local         
    ##  grDevices  * 3.5.0   2018-05-02 local         
    ##  grid         3.5.0   2018-05-02 local         
    ##  gtable       0.2.0   2016-02-26 CRAN (R 3.5.0)
    ##  htmltools    0.3.6   2017-04-28 CRAN (R 3.5.0)
    ##  knitr        1.20    2018-02-20 CRAN (R 3.5.0)
    ##  labeling     0.3     2014-08-23 CRAN (R 3.5.0)
    ##  lattice      0.20-35 2017-03-25 CRAN (R 3.5.0)
    ##  lazyeval     0.2.1   2017-10-29 CRAN (R 3.5.0)
    ##  magrittr     1.5     2014-11-22 CRAN (R 3.5.0)
    ##  memoise      1.1.0   2017-04-21 CRAN (R 3.4.2)
    ##  methods    * 3.5.0   2018-05-02 local         
    ##  munsell      0.4.3   2016-02-13 CRAN (R 3.5.0)
    ##  ncarbe     * 0.1.1   2018-06-07 local         
    ##  nlme         3.1-137 2018-04-07 CRAN (R 3.5.0)
    ##  NonCompart   0.4.2   2018-05-17 CRAN (R 3.5.0)
    ##  pillar       1.2.3   2018-05-25 CRAN (R 3.5.0)
    ##  pkgconfig    2.0.1   2017-03-21 CRAN (R 3.5.0)
    ##  plyr         1.8.4   2016-06-08 CRAN (R 3.5.0)
    ##  purrr        0.2.5   2018-05-29 CRAN (R 3.5.0)
    ##  R6           2.2.2   2017-06-17 CRAN (R 3.5.0)
    ##  Rcpp         0.12.17 2018-05-18 CRAN (R 3.5.0)
    ##  rlang        0.2.1   2018-05-30 CRAN (R 3.5.0)
    ##  rmarkdown    1.9     2018-03-01 CRAN (R 3.5.0)
    ##  rprojroot    1.3-2   2018-01-03 CRAN (R 3.5.0)
    ##  scales       0.5.0   2017-08-24 CRAN (R 3.5.0)
    ##  stats      * 3.5.0   2018-05-02 local         
    ##  stringi      1.2.2   2018-05-02 CRAN (R 3.5.0)
    ##  stringr      1.3.1   2018-05-10 CRAN (R 3.5.0)
    ##  tibble       1.4.2   2018-01-22 CRAN (R 3.5.0)
    ##  tidyselect   0.2.4   2018-02-26 CRAN (R 3.5.0)
    ##  tools        3.5.0   2018-05-02 local         
    ##  utf8         1.1.4   2018-05-24 CRAN (R 3.5.0)
    ##  utils      * 3.5.0   2018-05-02 local         
    ##  withr        2.1.2   2018-03-15 CRAN (R 3.5.0)
    ##  yaml         2.1.19  2018-05-01 CRAN (R 3.5.0)


NEWS.md
# ncarbe 0.1.1

- updated according to BELIB3.R

# ncarbe 0.1.0

- initial package



tests/test.R
library(tidyverse)

concData <- read.csv('inst/example/beConc.csv', stringsAsFactors = FALSE) %>%
  rename('Period' = 'PRD') %>% filter(SUBJ %in% c(5:10))



head(concData)

#PRD = 'Period'

ncarbe::beNCA(concData, PRD = 'Period')

# sas



dataBioeq <- read_csv('beNCAdataset.csv') %>%
  print()

model.Cmax <- nlme::lme(fixed = log(Cmax) ~ TRT + PRD + GRP,
                        random = ~1|SUBJ,
                        data = dataBioeq)

ci.Cmax <- nlme::intervals(model.Cmax, level=0.9, which="fixed")

model.AUClast <- nlme::lme(fixed = log(AUClast) ~ TRT + PRD + GRP,
                           random = ~1|SUBJ,
                           data = dataBioeq)

ci.AUClast <- nlme::intervals(model.AUClast, level=0.9, which="fixed")

confidenceInterval <- bind_rows(exp(ci.Cmax$fixed["TRTT", ]),
                                exp(ci.AUClast$fixed["TRTT", ])) %>%
  mutate(parameter = c('Cmax', 'AUClast')) %>%
  select(parameter,
         `Lower limit of 90% CI` = lower,
         `T/R ratio` = est.,
         `Upper limit of 90% CI` = upper)

ncarbestat <- list(Data = as.data.frame(dataBioeq),
                      Cmax = model.Cmax,
                      Summary.Cmax = summary(model.Cmax),
                      AUClast = model.AUClast,
                      Summary.AUClast = summary(model.AUClast),
                      Confidence.Interval = confidenceInterval)
print(ncarbestat)

#capture.output(yyNca$tadStat, file = 'docs/reports/nlme-report-Tad.txt')


R/beNCA.R
#' Performs a statistical analysis of 2x2 bioequivalence study
#'
#' \code{beNCA} returns text output of a statistical analysis of 2x2 bioequivalence study. Analysis of variance, least square means, confidence interval, and sample size will be calculated for AUClast and Cmax. Wilcoxon Signed-Rank Test and Hodges-Lehmann Estimate will be performed for Tmax.
#'
#' Reference: Shein-Chung Chow, Jen-pei Liu. Design and Analysis of Bioavailability and Bioequivalence Studies, 3rd ed. 2008. (ISBN:9781584886686)
#'
#' @param SUBJ Subject ID, any data type
#' @param GRP column name in which information of "RT" or "TR" exists.
#' @param PRD column name in which information of 1 or 2 exists.
#' @param TRT column name in which information of "R" or "T" exists.
#' @param method \code{kbe} by authors or \code{nlme} package uploaded on CRAN
#' @return returns text results of statistical analysis of 2x2 bioequivalence study including and \code{beNCAdataset.csv} file in the working directory which can be run in SAS.
#' @import NonCompart
#' @import dplyr
#' @importFrom nlme intervals lme
#' @export
#' @examples
#' file <- system.file('example', 'beConc.csv', package = 'ncarbe')
#' concData <- read.csv(file, as.is = TRUE)
#' beNCA(concData)

beNCA <- function(concData, SUBJ = 'SUBJ', GRP = 'GRP', PRD = 'PRD', TRT = 'TRT', method = 'kbe', ...){
  # SUBJ = 'SUBJ'; GRP = 'GRP';  PRD = 'PRD'; TRT = 'TRT'

  betestKey <- c('SUBJ', 'GRP', 'PRD', 'TRT')
  ncaKey <- setNames(c(SUBJ, GRP, PRD, TRT), betestKey)
  colnames(concData) <- sub(ncaKey['SUBJ'], 'SUBJ', colnames(concData))
  colnames(concData) <- sub(ncaKey['GRP'], 'GRP', colnames(concData))
  colnames(concData) <- sub(ncaKey['PRD'], 'PRD', colnames(concData))
  colnames(concData) <- sub(ncaKey['TRT'], 'TRT', colnames(concData))
  colnames(concData)

  bedataRaw <- NonCompart::tblNCA(as.data.frame(concData),
                                  key= betestKey,
                                  colTime="TIME", colConc="CONC", dose=100000, ..., R2ADJ = 0.5) %>%
    as.data.frame() %>%
    mutate_at(vars(AUCLST, CMAX, TMAX), as.numeric) %>%
    arrange(GRP, PRD, SUBJ) %>%
    select(SUBJ, TRT, GRP, PRD, AUClast = AUCLST, Cmax = CMAX, Tmax = TMAX)

  bedata <- bedataRaw %>%
    mutate(lnAUClast = log(AUClast),
           lnCmax = log(Cmax)) %>%
    as.data.frame()

  write.csv(bedataRaw, 'beNCAdataset.csv', row.names = FALSE)

  if(!assert(bedata)) {
    cat("\n Bad Data Format !\n");
    return(NULL);
  }

  if (method == 'kbe'){
    cat("\n\n[AUClast]\n\n");
    print(betest(bedata, "lnAUClast", logtransformed=TRUE));
    cat("\n\n[Cmax]\n\n");
    print(betest(bedata, "lnCmax", logtransformed=TRUE));
    cat("\n\n[Tmax]\n\n");
    print(hodges(bedata, "Tmax"))
  } else if (method == 'nlme') {

    dataBioeq <- bedata

    model.Cmax <- nlme::lme(fixed = log(Cmax) ~ TRT + PRD + GRP,
                            random = ~1|SUBJ,
                            data = dataBioeq)

    ci.Cmax <- nlme::intervals(model.Cmax, level=0.9, which="fixed")

    model.AUClast <- nlme::lme(fixed = log(AUClast) ~ TRT + PRD + GRP,
                               random = ~1|SUBJ,
                               data = dataBioeq)

    ci.AUClast <- nlme::intervals(model.AUClast, level=0.9, which="fixed")

    confidenceInterval <- bind_rows(exp(ci.Cmax$fixed["TRTT", ]),
                                    exp(ci.AUClast$fixed["TRTT", ])) %>%
      mutate(parameter = c('Cmax', 'AUClast')) %>%
      select(parameter,
             `Lower limit of 90% CI` = lower,
             `T/R ratio` = est.,
             `Upper limit of 90% CI` = upper)

    # ncarbestat <- list(Data = as.data.frame(dataBioeq),
    #                    Cmax = model.Cmax,
    #                    Summary.Cmax = summary(model.Cmax),
    #                    AUClast = model.AUClast,
    #                    Summary.AUClast = summary(model.AUClast),
    #                    Confidence.Interval = confidenceInterval)
    ncarbestat <- list(Confidence.Interval = confidenceInterval)
    print(ncarbestat)
  }
}



R-old/temp.R

print(betest(bedata, "lnAUClast", logtransformed=T))





bedata = bedata[order(bedata$GRP, bedata$PRD, bedata$SUBJ),];
if(!assert(bedata)) {
  cat("\n Bad Data Format !\n");
  return(NULL);
}



# misc ----


ncares[c(ncaKey, 'lnAUClast', 'lnCmax', 'Tmax')]


bedata$lnAUClast = log(bedata$AUClast);
bedata$lnCmax = log(bedata$Cmax);


kbe()

left_join(as_tibble(ncares) %>% gather(PPTESTCD, PPORRES, ),
          tibble(PPTESTCD = attr(ncares, 'dimnames')[[2]], UNIT = attr(ncares, 'units')),
          by = 'PPTESTCD') %>% 
  arrange(Subject, PPTESTCD) %>% 
  head(20)


bedata$lnAUClast = log(bedata$AUClast);
bedata$lnCmax = log(bedata$Cmax);

cat("\n\n[AUClast]\n\n");
print(betest(bedata, "lnAUClast", logtransformed=T));




library(tidyverse)
Theoph %>% 
  as_tibble() %>% 
  mutate_at(vars(Subject), as.numeric) %>% 
  select(Subject, Time, conc) %>% 
  arrange(Subject)

NonCompart::tblNCA()
function(concData = Theoph, PRD, TRT, GRP, key = , ...){
  
}
  
read.csv()
  
tblNCA(concData, key = "Subject", colTime = "Time", colConc = "conc", dose = 0, 
         adm = "Extravascular", dur = 0, doseUnit = "mg", timeUnit = "h", 
         concUnit = "ug/L", down = "Linear", MW = 0)

R/beFunctions.R
assert = function(bedata)
{
  Si11 = bedata[bedata$GRP=="RT" & bedata$PRD==1, "SUBJ"]
  Si21 = bedata[bedata$GRP=="RT" & bedata$PRD==2, "SUBJ"]
  Si12 = bedata[bedata$GRP=="TR" & bedata$PRD==1, "SUBJ"]
  Si22 = bedata[bedata$GRP=="TR" & bedata$PRD==2, "SUBJ"]

  return(identical(Si11, Si21) & identical(Si12, Si22))
}

betest = function(bedata, var, logtransformed)
{
# GRP, PRD
  if(!assert(bedata)) {
    cat("\n Bad Data Format !\n");
    return(NULL);
  }

  Yijk = bedata[, var]
  Yi11 = bedata[bedata$GRP=="RT" & bedata$PRD==1, var]
  Yi21 = bedata[bedata$GRP=="RT" & bedata$PRD==2, var]
  Yi12 = bedata[bedata$GRP=="TR" & bedata$PRD==1, var]
  Yi22 = bedata[bedata$GRP=="TR" & bedata$PRD==2, var]

  n1 = length(Yi11)
  n2 = length(Yi12)

  Y... = mean(Yijk)
  SStotal = sum((Yijk - Y...)^2)

  Y.11 = mean(Yi11)
  Y.21 = mean(Yi21)
  Y.12 = mean(Yi12)
  Y.22 = mean(Yi22)

  Yi.1 = (Yi11 + Yi21) / 2
  Yi.2 = (Yi12 + Yi22) / 2

  Y..1 = mean(Yi.1)
  Y..2 = mean(Yi.2)

  mu.r = (Y.11 + Y.22) / 2
  mu.t = (Y.21 + Y.12) / 2

  di1 = (Yi21 - Yi11) / 2
  di2 = (Yi22 - Yi12) / 2

  d.1 = mean(di1)
  d.2 = mean(di2)

#  sig2d = 1/(n1 + n2 - 2)*(sum((di1 - d.1)^2) + sum((di2 - d.2)^2))

#  Chat = Y.12 + Y.22 - Y.11 - Y.21
#  Fhat = mu.t - mu.r
#  Phat = (Y.21 - Y.11 - Y.12 + Y.22)/2

  SScarry   = 2*n1*n2/(n1 + n2)*(Y.12 + Y.22 - Y.11 - Y.21)^2 / 4
  SSinter   = (sum((Yi.1 - Y..1)^2) + sum((Yi.2 - Y..2)^2)) * 2
  SSbetween = SScarry + SSinter

  SSperiod  = 2*n1*n2/(n1+n2)*(Y.21 + Y.22 - Y.11 - Y.12)^2 / 4
  SSdrug    = 2*n1*n2/(n1+n2)*(Y.21 + Y.12 - Y.11 - Y.22)^2 / 4

  SSintra  = 2*(sum((di1 - d.1)^2) + sum((di2 - d.2)^2))
#  SSmodel = SStotal - SSintra

  Source = c("SUBJECT", "GROUP", "SUBJECT(GROUP)", "PERIOD", "DRUG", "ERROR", "TOTAL");
  SS     = c(SSbetween, SScarry, SSinter, SSperiod, SSdrug, SSintra, SStotal);
  DF     = c(n1 + n2 - 1, 1, n1 + n2 - 2, 1, 1, n1 + n2 - 2, 2*n1 + 2*n2 - 1);
  MS     = SS / DF
  mse    = SSintra/(n1 + n2 - 2)
  F      = MS / c(mse, MS[3], mse, mse, mse, mse, mse);
  p1 = 1 - pf(F[1], n1 + n2 - 1, n1 + n2 - 2)
  p2 = 1 - pf(F[2], 1, n1 + n2 - 2);
  p3 = 1 - pf(F[3], n1 + n2 - 2, n1 + n2 - 2);
  p4 = 1 - pf(F[4], 1, n1 + n2 - 2);
  p5 = 1 - pf(F[5], 1, n1 + n2 - 2);
  p  = c(p1, p2, p3, p4, p5, NA, NA)
  F[6] = F[7] = MS[7] = NA

  ANOVA = cbind(SS, DF, MS, F, p)
  dimnames(ANOVA) = list(Source,c("SS", "DF", "MS", "F", "p"))

  pe = mu.t - mu.r
  se = sqrt(mse / 2 * (1/n1 + 1/n2))   # See pp 62-63
  t0 = qt(0.95, n1 + n2 - 2);
  ci0 = cbind(pe - t0 * se, pe, pe + t0 * se)

  sig2b = (MS[3] - MS[6])/2
  sig2w = MS[6]

  if (logtransformed == TRUE) {
    cvs = cbind(sqrt(exp(sig2b) - 1), sqrt(exp(sig2w) - 1)) * 100

    lsm = cbind(exp(mu.r), exp(mu.t))
    dimnames(lsm) = list("Geometric Means", cbind("Reference Drug", "Test Drug"))

    ci = exp(ci0);
    dimnames(ci) = list("90% CI for Ratio", c("Lower Limit", "Point Estimate", "Upper Limit"));

    sampsize1 = bss.r.mse(mse);
    sampsize2 = bss.r.mse(mse, true.r=exp(pe));
    ss = cbind(sampsize1, sampsize2)
    dimnames(ss) = list("80% Power Sample Size", c("True Ratio=1", "True Ratio=Point Estimate"));
  } else {
    cvs = cbind(sqrt(sig2b)/mu.r, sqrt(sig2w)/mu.r) * 100

    lsm = cbind(mu.r, mu.t)
    dimnames(lsm) = list("Arithmetic Means", cbind("Reference Drug", "Test Drug"))

    ci1 = (1 + ci0 / mu.r) * 100
    ci = rbind(ci0, ci1)
    dimnames(ci) = list(c("90% CI for Difference", "90% CI for Difference(%)"), c("Lower Limit", "Point Estimate", "Upper Limit"));

    sampsize1 = bss.d.mse(mu.r, mse);
    sampsize2 = bss.d.mse(mu.r, mse, true.d=pe);
    ss = cbind(sampsize1, sampsize2)
    dimnames(ss) = list("80% Power Sample Size", c("True Difference=0", "True Difference=Point Estimate"));
  }
  cvs = rbind(cbind(sig2b, sig2w), cvs)
  dimnames(cvs) = list(cbind("Variance Estimate", "Coefficient of Variation, CV(%)"), cbind("Between Subject", "Within Subject"))

  result = list(ANOVA, cvs, lsm, ci, ss);
  names(result) = c("Analysis of Variance", "Between and Within Subject Variability", "Least Square Means", "90% Confidence Interval", "Sample Size")

  return(result);
}

hodges = function(bedata, var)
{
  if(!assert(bedata)) {
    cat("\n Bad Data Format !\n");
    return(NULL);
  }

  Yi11 = bedata[bedata$GRP=="RT" & bedata$PRD==1, var]
  Yi21 = bedata[bedata$GRP=="RT" & bedata$PRD==2, var]
  Yi12 = bedata[bedata$GRP=="TR" & bedata$PRD==1, var]
  Yi22 = bedata[bedata$GRP=="TR" & bedata$PRD==2, var]

  n1 = length(Yi11)
  n2 = length(Yi12)

  if(n1 * n2 < 12) {
    cat("\n Too Small Sample Size for 90% Confidence Interval !\n");
    return(NULL);
  }

  mu.r = (mean(Yi11) + mean(Yi22)) / 2;

  G1D = (Yi21 - Yi11) / 2
  G2D = (Yi22 - Yi12) / 2
  D = sort(outer(G1D, G2D, "-"));

  pval = pwilcox(min(length(D[D>0]), length(D[D<0])), n1, n2)
  w05 = qwilcox(0.05, n1, n2)
  w95 = qwilcox(0.95, n1, n2)

  names(pval) = list(c("p-value"));

  est1 = cbind(D[w05 - 1], median(D), D[w95])
  est2 = (1 + est1 / mu.r ) * 100
  est.a = rbind(est1, est2)
  dimnames(est.a) = list(c("90% Confidence Interval", "90% Confidence Interval(%)"), c("Lower Limit", "Point Estimate", "Upper Limit"));

#  est3 = cbind(D[w05], median(D), D[w95+1])
#  est4 = (1 + est3 / mu.r ) * 100
#  est.b = rbind(est3, est4)
#  dimnames(est.b) = list(c("90% Confidence Interval", "90% Confidence Interval(%)"), c("Lower Limit", "Point Estimate", "Upper Limit"));

#  result = list(pval, est.a, est.b);
#  names(result) = c("Wilcoxon Signed-Rank Test", "Hodges-Lehmann Estimate", "Hodges-Lehmann Estimate Old")

  result = list(pval, est.a);
  names(result) = c("Wilcoxon Signed-Rank Test", "Hodges-Lehmann Estimate")
  return(result);
}


bss.r.mse = function(mse, true.r=1, alpha=0.1, beta=0.2, thetaL=0.8, thetaU=1.25, nmax=999999)
{
  if (true.r <= thetaL | true.r >= thetaU) return(Inf)
  for (i in 2:nmax) {
    power = bpow.r.mse(i, i, mse, true.r, alpha, thetaL, thetaU)
    if (power > 1 - beta) return(i)
  }
  return(paste(">", nmax))
}

bpow.r.mse = function(n1, n2, mse, true.r=1, alpha=0.1, thetaL=0.8, thetaU=1.25)
{
# n1, n2: sample size for each group
  if (true.r <= thetaL | true.r >= thetaU) return(0)
  t0 = qt(1 - alpha/2, n1 + n2 - 2)
  p1 = pt(-1*t0, n1 + n2 - 2, ncp = log(thetaL/true.r) / sqrt(mse/2*(1/n1 + 1/n2)) ) # 1st beta error: LL > 0.8 = 1 - p1
  p2 = pt(t0, n1 + n2 - 2, ncp = log(thetaU/true.r) / sqrt(mse/2*(1/n1 + 1/n2)) )    # 2nd beta error: UL < 1.25 = p2
  power = p1 - p2  # 1 - (1 - p1 + p2) = 1 - both beta errors
  if (power < 0) power = 0
  return(power)
}
